src/main/java/org/opentripplanner/middleware/controllers/api/ApiKeyManagementController.java

```diff
@@ -0,0 +1,150 @@
+        if (apiKey == null) {
+            logMessageAndHalt(req,
+                HttpStatus.INTERNAL_SERVER_ERROR_500,
+                String.format("Unable to get AWS api key for user (%s)", user),
+                null);
+        }
+
+        // add new api key to user and persist
+        user.apiKeyIds.add(apiKey.getId());
+        apiUsers.replace(user.id, user);
+
+        // FIXME assumption is that the api key will be required by the calling service
+        return apiKey;
+    }
+
+    /**
+     * Delete an api key from a given user's list of api keys (if present) and from AWS api gateway.
+     */
+    private static boolean deleteApiKeyForApiUser(Request req, Response res) {
+
+        ApiUser user = getApiUser(req);
+
+        String apiKeyId = req.queryParamOrDefault("apiKeyId", null);
+        if (apiKeyId == null) {
+            logMessageAndHalt(req,
+                HttpStatus.BAD_REQUEST_400,
+                "An api key id is required",
+                null);
+        }
+
+        // FIXME perhaps not required?
+        if (!user.apiKeyIds.contains(apiKeyId)) {
+            logMessageAndHalt(req,
+                HttpStatus.BAD_REQUEST_400,
+                String.format("User id (%s) does not have expected api key id (%s)", user.id, apiKeyId),
+                null);
+        }
+
+        // delete api key from user and persist
+        user.apiKeyIds.remove(apiKeyId);
+        apiUsers.replace(user.id, user);
+
+        // delete api key from AWS
+        ApiGatewayUtils.deleteApiKeys(Collections.singletonList(apiKeyId));
+
+        return true;
+    }
+
+    /**
+     * Get API usage from AWS api gateway for a users api keys.
+     */
+    private static List<GetUsageResult> getUsageLogsForApiUser(Request req, Response res) {
+
+        ApiUser user = getApiUser(req);
+
+        LocalDateTime now = LocalDateTime.now();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        String startDate = req.queryParamOrDefault("startDate", formatter.format(now.minusDays(30)));
+        String endDate = req.queryParamOrDefault("endDate", formatter.format(now));
+
+        try {
+            // FIXME the GetUsageResult does not include the api key id. If the api key id is required the response will
+            //  need to be enhanced.
+            return ApiGatewayUtils.getUsageLogs(user.apiKeyIds, startDate, endDate);
+        } catch (Exception e) {
+            // Catch any issues with bad request parameters (e.g. bad date format).
+            logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Error requesting usage results", e);
+        }
+
+        return null;
+    }
+
+    /**
+     * Get an Api user from Mongo DB based on the provided user id. If no user id is provided, or no matching Api user
+     * can be found, respond and halt.
+     */
+    private static ApiUser getApiUser(Request req) {
+
+        String userId = req.queryParamOrDefault("userId", null);
```

Comment:
User Id should be an optional query param. ApiUser should be derived from token if the param is not provided (assumption is that the request is coming from the ApiUser themselves in that case).

Reply:
Updated.